### PR TITLE
Add Von Mises Metric for Structural Mechanics Applications

### PR DESCRIPTION
**Pull request by mnabian**
# PhysicsNeMo Pull Request

## Description

Add von Mises stress metric for structural mechanics. This will be used in the structural mechanics recipes.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: physicsnemo/metrics/cae/von_mises.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+"""
+
+from typing import Union
+
+import numpy as np
+import torch
+
+
+def von_mises_from_stress_tensor(
+    stress: Union[torch.Tensor, np.ndarray],
```
Comment (peterdsharpe) [inline physicsnemo/metrics/cae/von_mises.py]: In new code, we should always prefer modern `|` union types over `Union`. (Both here and throughout)

File: physicsnemo/metrics/cae/von_mises.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+) -> Union[torch.Tensor, np.ndarray]:
```
Comment (peterdsharpe) [inline physicsnemo/metrics/cae/von_mises.py]: These should all use `jaxtyping` to be compliant with repo rules. (Both here and throughout, on all type signatures for tensors)

File: physicsnemo/metrics/cae/von_mises.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        # Symmetrize shear components for robustness to slight asymmetry
+        sxy = 0.5 * (stress[..., 0, 1] + stress[..., 1, 0])
+        syz = 0.5 * (stress[..., 1, 2] + stress[..., 2, 1])
+        sxz = 0.5 * (stress[..., 0, 2] + stress[..., 2, 0])
+        vm_sq = 0.5 * ((sxx - syy) ** 2 + (syy - szz) ** 2 + (szz - sxx) ** 2) + 3.0 * (
+            sxy**2 + syz**2 + sxz**2
+        )
+        return torch.sqrt(torch.clamp(vm_sq, min=0.0))
+
+    _validate_stress_shape(np.shape(stress))
```
Comment (peterdsharpe) [inline physicsnemo/metrics/cae/von_mises.py]: For readability, and for better error handling, I'd recommend always prefixing switch-statement clauses with an:

```python
elif isinstance(stress, np.ndarray):
```

And then having an:

```python
else:
    raise TypeError("...")
```

In other words, switch statements should always declare their switches.

File: physicsnemo/metrics/cae/von_mises.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    sxy = 0.5 * (stress[..., 0, 1] + stress[..., 1, 0])
+    syz = 0.5 * (stress[..., 1, 2] + stress[..., 2, 1])
+    sxz = 0.5 * (stress[..., 0, 2] + stress[..., 2, 0])
+    vm_sq = 0.5 * ((sxx - syy) ** 2 + (syy - szz) ** 2 + (szz - sxx) ** 2) + 3.0 * (
+        sxy**2 + syz**2 + sxz**2
+    )
+    return np.sqrt(np.maximum(vm_sq, 0.0))
+
+
+def _validate_stress_shape(shape):
```
Comment (peterdsharpe) [inline physicsnemo/metrics/cae/von_mises.py]: Missing type signature here

File: physicsnemo/metrics/cae/von_mises.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    stress = np.asarray(stress, dtype=np.float64)
+    sxx = stress[..., 0, 0]
+    syy = stress[..., 1, 1]
+    szz = stress[..., 2, 2]
+    sxy = 0.5 * (stress[..., 0, 1] + stress[..., 1, 0])
+    syz = 0.5 * (stress[..., 1, 2] + stress[..., 2, 1])
+    sxz = 0.5 * (stress[..., 0, 2] + stress[..., 2, 0])
+    vm_sq = 0.5 * ((sxx - syy) ** 2 + (syy - szz) ** 2 + (szz - sxx) ** 2) + 3.0 * (
+        sxy**2 + syz**2 + sxz**2
+    )
```
Comment (peterdsharpe) [inline physicsnemo/metrics/cae/von_mises.py]: This can be vectorized, if desired, using `einsum`. This may yield performance benefits if this is used downstream in performance-critical applications - up to you.

It's also possible to write this code in a way that's dimensionally-generic (e.g., works on 2x2 stress tensors for 2D problems), which could be nice down the road for reusability?

File: physicsnemo/metrics/cae/von_mises.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    >>> # Uniaxial tension: σxx = 100 MPa, all others zero
+    >>> stress = torch.tensor([[100., 0., 0.], [0., 0., 0.], [0., 0., 0.]])
+    >>> von_mises_from_stress_tensor(stress)
+    tensor(100.)
+    >>> # Hydrostatic stress (no distortion): σxx = σyy = σzz = 50
+    >>> stress = torch.eye(3) * 50
+    >>> von_mises_from_stress_tensor(stress)
+    tensor(0.)
+    """
+    if isinstance(stress, torch.Tensor):
```
Comment (peterdsharpe) [inline physicsnemo/metrics/cae/von_mises.py]: It might be worth using the Array API standard to deduplicate the parallel numpy and torch code here? This is the official effort coordinated across NumPy/Torch/JAX/others to allow for multiple dispatch on arrays.

https://data-apis.org/array-api-compat/

I've used this in other projects - it's excellent and makes for very clean reusable implementations.

File: CHANGELOG.md [hunk 7]
```diff
@@ -18,6 +18,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   `.to_dual_graph()` methods. These allow Mesh conversion to 0D point clouds, 1D
   edge graphs, and 1D dual graphs, respectively, when connectivity information
   is not needed.
+- Adds von Mises stress (`physicsnemo.metrics.cae.von_mises`).
```
Comment (peterdsharpe) [inline CHANGELOG.md]:
```suggestion
- Adds von Mises stress (`physicsnemo.metrics.cae.von_mises`) as a CAE metric.
```

**Review by peterdsharpe — CHANGES_REQUESTED** [review]
A few changes requested, but overall this'll be a nice new feature when merged.
